snuba/datasets/errors_replacer.py

```diff
@@ -761,9 +761,25 @@ def process_unmerge_hierarchical(
         ),
     }
 
-    # Sentry is expected to send an `exclude_groups` message after unsplit is
-    # done, and we can live with data inconsistencies while this is ongoing.
-    query_time_flags = (None, message["project_id"])
+    if message.get("skip_needs_final", False):
+        # In some cases Sentry can ensure that no events remain in
+        # `previous_group_id`. In that case FINAL is not applied to future
+        # queries, Sentry is instead expected to send an `exclude_groups`
+        # message after unsplit is done, and we can live with data
+        # inconsistencies while this is ongoing.
```

Comment:
Could you please expand on when we would do a split that leaves events in the previous group ? Adding final is a huge performance issue for a project. On a big project it makes it barely usable for 24 hours, so far we could get away with that only because unmerge is a quite rare operation. For split we cannot get away with that.

Reply:
We do *not* need final on split. We need final on *un*-split: The reverse action where the grouping level is reduced, i.e. user regrets their choice. This is where I draw the analogy to unmerge: It is not a common operation because it's undoing a user action.

Why do we need FINAL? Because when undoing a split, we create a new group that draws in events from other issues. Those issues may be in a merged state, so they may contain unrelated events.

I see ways to get rid of FINAL requirement: We could first perform a regular unmerge on those particular issues. Then unmerge will trigger FINAL, but I think there is a way to get rid of FINAL for unmerge as well, with a little trickery on Sentry side.

Reply:
Just to clarify: I am choosing to take a shortcut here with NEEDS_FINAL such that we can get to a functional state. The reason is the same as for unmerge. If this becomes a practical performance concern I see ways to eliminate the cost, but it requires significant work in Sentry such that neither unmerge nor unmerge_hierarchical ever leaves events in `old_group_id`.

Reply:
So I think there are two areas on which we are not on the same page.

My understanding from the discussions we had last months on the state chart was that if you merge two groups with different primary hashes, a split/unsplit that involved a merged group at that point would be forbidden. This was my understanding from the edges/states on the top right of the diagram. 

So when splitting a group you would be rewriting all events for that group and same it would be when unsplitting. 
Has this changed since back then ?

Also as a consequence, I thought you would be using this method (unmerge_hierarchical) for split and not unsplit, as for unsplit you would be using `merge`. 
Now if this becomes a generic "rewrite_some_events_with_a_given_hierarchical_hash" I would rename into "rewrite_hierarchical" or something similar as it is not meant only to split/unmerge events anymore.

Reply:
You are totally right. For now I'll revert these changes, I think we can live with this restriction.

In my mind the restriction only existed for split, but if it exists for unsplit too that's fine as well. The latter is just more difficult to check for.